server/src/gnome/prompter.rs

```diff
@@ -0,0 +1,205 @@
+                        Value::new("Lock Keyring").try_to_owned().unwrap(),
+                    );
+                    properties.insert("caller-window", Value::new("").try_to_owned().unwrap());
+                    properties.insert("cancel-label", Value::new("Cancel").try_to_owned().unwrap());
+
+                    let connection = connection.clone();
+                    let path = self.path.clone();
+
+                    tokio::spawn(PrompterCallback::perform_prompt(
+                        connection, path, "confirm", properties, exchange,
+                    ));
+                } else if reply == "no" {
+                    // Second PromptReady call and the prompt is dismissed
+                    tracing::debug!("Prompt is being dismissed.");
+
+                    tokio::spawn(PrompterCallback::stop_prompting(
+                        connection.clone(),
+                        self.path.clone(),
+                    ));
+
+                    let signal_emitter = self.service.signal_emitter(prompt.path().clone())?;
+                    let result = Value::new::<Vec<OwnedObjectPath>>(vec![])
+                        .try_to_owned()
+                        .unwrap();
+
+                    tokio::spawn(PrompterCallback::prompt_completed(
+                        signal_emitter,
+                        true,
+                        result,
+                    ));
+                } else {
+                    // Second PromptReady call with the final exchange
+                    let service = self.service.clone();
+                    let objects = prompt.objects().clone();
+                    let result = Value::new(&objects).try_to_owned().unwrap();
+
+                    tokio::spawn(async move {
+                        let _ = service.set_locked(true, &objects, true).await;
+                    });
+
+                    tokio::spawn(PrompterCallback::stop_prompting(
+                        connection.clone(),
+                        self.path.clone(),
+                    ));
+
+                    let signal_emitter = self.service.signal_emitter(prompt.path().clone())?;
+
+                    tokio::spawn(PrompterCallback::prompt_completed(
+                        signal_emitter,
+                        false,
+                        result,
+                    ));
+                }
+            }
+        };
+
+        Ok(())
+    }
+
+    pub async fn prompt_done(
+        &self,
+        #[zbus(object_server)] object_server: &zbus::ObjectServer,
+    ) -> Result<(), ServiceError> {
+        if let Some(prompt) = self.service.prompt().await {
+            object_server.remove::<Prompt, _>(prompt.path()).await?;
+            self.service.remove_prompt().await;
+        }
+        object_server.remove::<Self, _>(&self.path).await?;
+
+        Ok(())
+    }
+}
+
+impl PrompterCallback {
+    pub async fn new(service: Service) -> Self {
+        let index = service.prompt_index().await;
```

Comment:
You should create a PrompterProxy here and keepe it around for the functions below.

Reply:
Good idea. But I can't access `self` inside `tokio::spawn`. All these following functions are associated functions.